src/StimulusBundle/src/Form/Extension/FormTypeExtension.php

```diff
@@ -0,0 +1,162 @@
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\UX\StimulusBundle\Dto\StimulusAttributes;
+use Twig\Environment;
+use Twig\Loader\ArrayLoader;
+
+class FormTypeExtension extends AbstractTypeExtension
+{
+    private StimulusAttributes $stimulusAttributes;
+
+    public static function getExtendedTypes(): iterable
+    {
+        return [FormType::class];
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options): void
+    {
+        if (
+            isset($options['stimulus_controller'])
+            || !isset($options['stimulus_target'])
+            || !isset($options['stimulus_action'])
+        ) {
+            $this->stimulusAttributes = new StimulusAttributes(new Environment(new ArrayLoader()));
+
+            if (isset($options['stimulus_controller'])) {
+                $this->handleController($options['stimulus_controller']);
+            }
+
+            if (isset($options['stimulus_target'])) {
+                $this->handleTarget($options['stimulus_target']);
+            }
+
+            if (isset($options['stimulus_action'])) {
+                $this->handleAction($options['stimulus_action']);
+            }
+
+            $attributes = array_merge($view->vars['attr'], $this->stimulusAttributes->toArray());
+            $view->vars['attr'] = $attributes;
+        }
+
+        foreach (['row_attr', 'choice_attr'] as $index) {
+            if (
+                isset($options[$index])
+                && (
+                    isset($options[$index]['stimulus_controller'])
+                    || isset($options[$index]['stimulus_target'])
+                    || isset($options[$index]['stimulus_action'])
+                )
+            ) {
+                $this->stimulusAttributes = new StimulusAttributes(new Environment(new ArrayLoader()));
+
+                if (isset($options[$index]['stimulus_controller'])) {
+                    $this->handleController($options[$index]['stimulus_controller']);
+                    unset($options[$index]['stimulus_controller']);
+                }
+
+                if (isset($options[$index]['stimulus_target'])) {
+                    $this->handleTarget($options[$index]['stimulus_target']);
+                    unset($options[$index]['stimulus_target']);
+                }
+
+                if (isset($options[$index]['stimulus_action'])) {
+                    $this->handleAction($options[$index]['stimulus_action']);
+                    unset($options[$index]['stimulus_action']);
+                }
+
+                $attributes = array_merge($options[$index], $this->stimulusAttributes->toArray());
+                $view->vars[$index] = $attributes;
+            }
+        }
+    }
+
+    private function handleController(string|array $controllers): void
+    {
+        if (\is_string($controllers)) {
+            $controllers = [$controllers];
```

Comment:
Could be handle outside this method i suppose (will call $this->stimulusAttributes->addController(...) every case)